NetworkRTCTCPSocketCocoa::getInterfaceName is not always answering its completion handler

youennf opened the pull request:
#### 877db403d4d96b65af264a22e5916087dceca591
<pre>
NetworkRTCTCPSocketCocoa::getInterfaceName is not always answering its completion handler
<a href="https://rdar.apple.com/151924020">rdar://151924020</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=293489">https://bugs.webkit.org/show_bug.cgi?id=293489</a>

Reviewed by Jean-Yves Avenard.

We migrate NetworkRTCTCPSocketCocoa::getInterfaceName to a NativePromise.
This simplifies both the thread management and the fact that the callback is always executed and only once.
To fix the issue of not always answering NetworkRTCTCPSocketCocoa::getInterfaceName completion handler, we use an auto reject promise producer.
Minor refactoring to remove protectedRTCNetworkThreadQueue given we can use a const Ref for the work queue.

* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.cpp:
(WebKit::NetworkRTCProvider::getInterfaceName):
(WebKit::NetworkRTCProvider::callOnRTCNetworkThread):
(WebKit::NetworkRTCProvider::assertIsRTCNetworkThread):
(WebKit::NetworkRTCProvider::protectedRTCNetworkThreadQueue): Deleted.
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
(WebKit::NetworkRTCTCPSocketCocoa::getInterfaceName):

Canonical link: <a href="https://commits.webkit.org/295415@main">https://commits.webkit.org/295415@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42f2c69f663590d5af49deaa13aec0b236417cda

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105025 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110242 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55705 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25149 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33283 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79755 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108031 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19575 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94793 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60062 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19327 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55082 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89043 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12914 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112745 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32190 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23693 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88836 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32556 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91013 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88465 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22552 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33355 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11143 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27559 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32114 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37487 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31907 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35248 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33466 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6741fc7dd0b7e7ac9734a345cec802ceffdf40bd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45835

youennf commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.h:
```diff
@@ -29,6 +29,7 @@
 
 #include "NetworkRTCProvider.h"
 #include <Network/Network.h>
+#include <wtf/NativePromise.h>
```
done

youennf commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
```diff
@@ -223,24 +223,19 @@ static inline void processIncomingData(RetainPtr<nw_connection_t>&& nwConnection
 
     nw_connection_set_queue(nwConnection.get(), tcpSocketQueue());
-    nw_connection_set_state_changed_handler(nwConnection.get(), makeBlockPtr([callback = WTFMove(callback), nwConnection](nw_connection_state_t state, _Nullable nw_error_t error) mutable {
+    nw_connection_set_state_changed_handler(nwConnection.get(), makeBlockPtr([promiseProducer = WTFMove(promiseProducer), nwConnection](nw_connection_state_t state, _Nullable nw_error_t error) mutable {
```
It can and it will. As soon as we cal resolve/reject, we are setting nwConnection to nullptr and we check nwConnection before resolving/rejecting so normally we should be ok.

youennf committed:
NetworkRTCTCPSocketCocoa::getInterfaceName is not always answering its completion handler
rdar://151924020
https://bugs.webkit.org/show_bug.cgi?id=293489

Reviewed by Jean-Yves Avenard.

We migrate NetworkRTCTCPSocketCocoa::getInterfaceName to a NativePromise.
This simplifies both the thread management and the fact that the callback is always executed and only once.
To fix the issue of not always answering NetworkRTCTCPSocketCocoa::getInterfaceName completion handler, we use an auto reject promise producer.
Minor refactoring to remove protectedRTCNetworkThreadQueue given we can use a const Ref for the work queue.

* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.cpp:
(WebKit::NetworkRTCProvider::getInterfaceName):
(WebKit::NetworkRTCProvider::callOnRTCNetworkThread):
(WebKit::NetworkRTCProvider::assertIsRTCNetworkThread):
(WebKit::NetworkRTCProvider::protectedRTCNetworkThreadQueue): Deleted.
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
(WebKit::NetworkRTCTCPSocketCocoa::getInterfaceName):

Canonical link: https://commits.webkit.org/295415@main
